### PR TITLE
wallet: Use CTxDestination in CRecipient instead of just scriptPubKey

### PR DESCRIPTION
**Pull request by achow101**
For silent payments, we want to provide a `SilentPaymentsDestination` to be used as the recipient, which requires `CRecipient` to use something other than just the `scriptPubKey` as we cannot know the output script for a silent payment prior to transaction creation. `CTxDestination` seems like the obvious place to add a `SilentPaymentsDestination` as it is our internal representation of an address.

In order to still allow paying to arbitrary scriptPubKeys (e.g. for data carrier outputs, or the user hand crafted a raw transaction that they have given to `fundrawtransaction`), `CNoDestination` is changed to contain raw scripts.

Additionally, P2PK scripts are now interpreted as a new `PubKeyDestination` rather than `PKHash`. This results in some things that would have given an address for P2PK scripts to no longer do so. This is arguably more correct.

`ExtractDestination`'s behavior is slightly changed for the above. It now returns `true` for those destinations that have addresses, so P2PK scripts now result in `false`. Even though it returns false for `CNoDestination`, the script will now be included in that `CNoDestination`.

Builds on #28244

### CODE REVIEW

File: src/addresstype.h [hunk 1]
```diff
@@ -23,6 +23,15 @@ class CNoDestination {
     friend bool operator<(const CNoDestination &a, const CNoDestination &b) { return a.script < b.script; }
 };
 
+struct PubKeyDestination
```
Comment (josibake) [inline src/addresstype.h:26]: in https://github.com/bitcoin/bitcoin/pull/28246/commits/34d7ab05b3f30e2827d9358a212f25865a71cce8:

Looks like this is causing the fuzzer to fail in `src/test/fuzz/util.cpp` because this new type is not handled.

Comment (achow101) [reply src/addresstype.h:26]: Should be fixed.

File: src/addresstype.cpp [hunk 2]
```diff
@@ -94,6 +94,7 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
     case TxoutType::MULTISIG:
     case TxoutType::NULL_DATA:
     case TxoutType::NONSTANDARD:
+        addressRet = CNoDestination(scriptPubKey);
```
Comment (S3RK) [inline src/addresstype.cpp:98]: Comment for `ExtractDestination` function needs to be updated.

File: src/wallet/spend.cpp [hunk 3]
```diff
@@ -391,8 +391,14 @@ std::map<CTxDestination, std::vector<COutput>> ListCoins(const CWallet& wallet)
     coins_params.skip_locked = false;
     for (const COutput& coin : AvailableCoins(wallet, &coin_control, /*feerate=*/std::nullopt, coins_params).All()) {
         CTxDestination address;
-        if ((coin.spendable || (wallet.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && coin.solvable)) &&
-            ExtractDestination(FindNonChangeParentOutput(wallet, coin.outpoint).scriptPubKey, address)) {
+        if ((coin.spendable || (wallet.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && coin.solvable))) {
+            if (!ExtractDestination(FindNonChangeParentOutput(wallet, coin.outpoint).scriptPubKey, address)) {
```
Comment (S3RK) [inline src/wallet/spend.cpp:395]: Maybe add a comment that this is done for backward compatibility?

File: src/rpc/output_script.cpp [hunk 4]
```diff
@@ -280,7 +280,10 @@ static RPCHelpMan deriveaddresses()
                 for (const CScript& script : scripts) {
                     CTxDestination dest;
                     if (!ExtractDestination(script, dest)) {
-                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Descriptor does not have a corresponding address");
+                        if (!std::get_if<PubKeyDestination>(&dest)) {
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/rpc/output_script.cpp:283]: oh, I see. It's because `combo()` descriptors generate P2PK.. ugh..

**Comment by S3RK** [comment]
CodeReview ACK b2a0586cae3650975f8f1a2048cdad4e0274a9d1 with some nits

**Comment by Sjors** [comment]
Can we add `CRawDestination` and keep `CNoDestination` for where we couldn't parse the destination? This seems safer, e.g. `AddressTableModel::setData` (GUI) checks for this.

**Comment by achow101** [comment]
> Can we add `CRawDestination` and keep `CNoDestination` for where we couldn't parse the destination?

That isn't a useful distinction. We can't parse the destination when it's just a random script, and a random script would be classified as a raw destination.

File: src/addresstype.cpp (hunk 2 above)
Comment (achow101) [reply src/addresstype.cpp:98]: Updated the comments

File: src/rpc/output_script.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/output_script.cpp]: Added a comment. Also noticed a minor bug with deriving a `pk()` descriptor, so added a fix and test for that.

File: src/wallet/spend.cpp (hunk 3 above)
Comment (achow101) [reply src/wallet/spend.cpp:395]: Done

**Comment by Sjors** [comment]
> We can't parse the destination when it's just a random script

But a random script is distinct from an invalid encoding. For the UI that different matters. (though maybe doesn't require a different CDestination thing)

> a random script would be classified as a raw destination

That sounds correct to me?

> Decoding into a raw script is also fundamentally impossible

How do you mean?

**Comment by achow101** [comment]
> > We can't parse the destination when it's just a random script
> 
> But a random script is distinct from an invalid encoding. For the UI that different matters. (though maybe doesn't require a different CDestination thing)

The UI is still getting a `CNoDestination` anyways. It's only decoding addresses (or things purporting to be addresses) which cannot be arbitrary scripts.

> > Decoding into a raw script is also fundamentally impossible
> 
> How do you mean?

There's no address type that maps to arbitrary scripts. It's literally impossible for `DecodeDestination` to return a `CNoDestination` that contains an arbitrary script since decoding fails first.

This PR makes it so that a `CNoDestination` continues to mean that when returned from `DecodeDestination`, there was failure to decode, and when returned from `ExtractDestination`, the provided script does not match any standard template. I think what you're proposing is to essentially break that up, but in that case, `CNoDestination` is basically just a boolean that would be better served as a `util::Result` as it has no meaning outside of that context.

**Comment by Sjors** [comment]
I'm thinking we might have a future where someone - who knows what they're doing - can paste in an arbitrary `scriptPubKey` in hex format. At least in the RPC, and maybe also in the GUI. E.g. because they want to send to a bare multisig. Or to some far future soft fork script that's not following the `bc1[version, rest of script]` segwit pattern (and for some reason they don't want to upgrade the wallet). This could then be stored as a `CRawDestination`.

> It's literally impossible for DecodeDestination to return a CNoDestination that contains an arbitrary script since decoding fails first.

> I think what you're proposing is to essentially break that up

I think I'm proposing a change to `DecodeDestination` so that it returns `CRawDestination` if it's a scriptPubKey and `CNoDestination` if it's really unparseable.

**Comment by Sjors** [comment]
That said, since we don't need `CRawDestination` for Silent Payments, maybe it's a bit overkill and we can introduce something like `CRawDestination` later.

**Comment by achow101** [comment]
> I think I'm proposing a change to `DecodeDestination` so that it returns `CRawDestination` if it's a scriptPubKey and `CNoDestination` if it's really unparseable.

That's way out of scope for this PR.

**Comment by josibake** [comment]
Concept ACK

In #28122 , I also need to update `fuzz/util.cpp` to handle the new `V0SilentPaymentsDestination` type. I realised we are now duplicating code in three places to create pubkeys for the fuzzer, so I pulled your code for creating compressed/uncompressed pubkeys into a function here: https://github.com/bitcoin/bitcoin/pull/28361. 

Can simplify this PR a bit by rebasing on top of https://github.com/bitcoin/bitcoin/pull/28361 (assuming you agree with the approach)

**Comment by josibake** [comment]
> Can we add `CRawDestination` and keep `CNoDestination` for where we couldn't parse the destination? This seems safer, e.g. `AddressTableModel::setData` (GUI) checks for this.
> 
> > Builds on #28244
> 
> That was merged.
> 
> The first commit triggers a `maybe-uninitialized` warning for me (Ubuntu 23.10 with gcc 12.3.0):
> 
> ```
> /prevector.h:170:37: error: ‘*(const prevector<28, unsigned char, unsigned int, int>*)((char*)&<unnamed> + offsetof(std::CTxDestination, std::variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Variant_base<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Move_assign_base<false, CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Copy_assign_base<false, CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Move_ctor_base<false, CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Copy_ctor_base<false, CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::<unnamed>.std::__detail::__variant::_Variant_storage<false, CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>::_M_u)).prevector<28, unsigned char>::_size’ may be used uninitialized [-Werror=maybe-uninitialized]
>   170 |     bool is_direct() const { return _size <= N; }
>       |                                     ^~~~~
> util/message.cpp: In function ‘MessageVerificationResult MessageVerify(const std::string&, const std::string&, const std::string&)’:
> util/message.cpp:50:40: note: ‘<anonymous>’ declared here
>    50 |     if (!(CTxDestination(PKHash(pubkey)) == destination)) {
> ```

Seeing this as well on the latest push

**Comment by achow101** [comment]
> The first commit triggers a `maybe-uninitialized` warning for me (Ubuntu 23.10 with gcc 12.3.0):

Unable to get this warning at all even with gcc 12.3.0

**Comment by josibake** [comment]
> Unable to get this warning at all even with gcc 12.3.0

Could be a false positive? But not sure how to confirm that

**Comment by instagibbs** [comment]
concept ACK, makes sense to ferry additional info around and use `CTxDestination` in wallet contexts like this

**Review by maflcko — COMMENTED** [review]
left some style nits, nothing blocking, feel free to ignore

File: src/addresstype.h [hunk 5]
```diff
@@ -69,22 +69,25 @@ struct WitnessV1Taproot : public XOnlyPubKey
 //! CTxDestination subtype to encode any future Witness version
 struct WitnessUnknown
 {
-    unsigned int version;
-    unsigned int length;
-    unsigned char program[40];
```
Comment (josibake) [inline src/addresstype.h:74]: AFAICT, changing this to `std::vector` was causing the `-Wuninitialized` warning. I played around with overriding the default constructor but kept getting the same error. What was causing the error was this line:

https://github.com/bitcoin/bitcoin/blob/e3d6bd41629a0735d7fb1475b77ffd3be7280c9a/src/util/message.cpp#L50

where we temporarily create a `CTxDestination(PKHash(pubkey))` to compare it to a `CTxDestination`. From what I can tell, creating the temporary `CTxDestination` with only a `PKHash` in it and then immediately calling the destructor was triggering the error for `WitnessUnknown` since it was never really constructed. Either way, I think it makes more sense to just create a `PKHash` and then compare it to the `CTxDestination` with a `std::get_if` like so: 

```
-    if (!(CTxDestination(PKHash(pubkey)) == destination)) {
+    if (!(PKHash(pubkey) == *std::get_if<PKHash>(&destination))) {
```

I made this change in https://github.com/bitcoin/bitcoin/commit/50fd2a5455288a87249ef5ed73f7db34f814bf8a and verified that everything compiles without the warning, feel free to cherry-pick it

File: src/test/fuzz/util.cpp [hunk 6]
```diff
@@ -163,6 +163,13 @@ CTxDestination ConsumeTxDestination(FuzzedDataProvider& fuzzed_data_provider) no
         [&] {
             tx_destination = CNoDestination{};
         },
+        [&] {
+            uint8_t pk_type = fuzzed_data_provider.PickValueInArray({0x02, 0x03, 0x04, 0x06, 0x07});
+            std::vector<uint8_t> pk_data = ConsumeFixedLengthByteVector(fuzzed_data_provider, pk_type >= 0x04 ? 65 : 33);
+            pk_data[0] = pk_type;
+            CPubKey pk{pk_data.begin(), pk_data.end()};
+            tx_destination = PubKeyDestination{pk};
+        },
```
Comment (josibake) [inline src/test/fuzz/util.cpp:172]: https://github.com/bitcoin/bitcoin/pull/28361 merged, so this can become:

```suggestion
        [&] {
            bool compressed = fuzzed_data_provider.ConsumeBool();
            CPubKey pk{ConstructPubKeyBytes(
                    fuzzed_data_provider,
                    ConsumeFixedLengthByteVector(fuzzed_data_provider, (compressed ? CPubKey::COMPRESSED_SIZE : CPubKey::SIZE)),
                    compressed
            )};
            tx_destination = PubKeyDestination{pk};
        },
```

File: src/addresstype.h [hunk 7]
```diff
@@ -69,22 +69,25 @@ struct WitnessV1Taproot : public XOnlyPubKey
...
+private:
+    unsigned int m_version;
+    std::vector<unsigned char> m_program;
+
+public:
+    WitnessUnknown(unsigned int version, std::vector<unsigned char> program) : m_version(version), m_program(program) {}
+
+    unsigned int GetWitnessVersion() const { return m_version; }
+    const std::vector<unsigned char>& GetWitnessProgram() const { return m_program; }
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h]: Done

File: src/rpc/util.cpp [hunk 8]
```diff
@@ -303,8 +303,8 @@ class DescribeAddressVisitor
     {
         UniValue obj(UniValue::VOBJ);
         obj.pushKV("iswitness", true);
-        obj.pushKV("witness_version", (int)id.version);
-        obj.pushKV("witness_program", HexStr({id.program, id.length}));
+        obj.pushKV("witness_version", (int)id.GetWitnessVersion());
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/util.cpp]: Done

File: src/test/script_standard_tests.cpp [hunk 9]
```diff
@@ -249,10 +249,7 @@ BOOST_AUTO_TEST_CASE(script_standard_ExtractDestination)
     s.clear();
     s << OP_1 << ToByteVector(pubkey);
     BOOST_CHECK(ExtractDestination(s, address));
-    WitnessUnknown unk;
-    unk.length = 33;
-    unk.version = 1;
-    std::copy(pubkey.begin(), pubkey.end(), unk.program);
+    WitnessUnknown unk(1, ToByteVector(pubkey));
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/test/script_standard_tests.cpp]: Done

File: src/addresstype.h [hunk 10]
```diff
@@ -69,22 +69,25 @@ struct WitnessV1Taproot : public XOnlyPubKey
...
 {
-    unsigned int version;
-    unsigned int length;
-    unsigned char program[40];
+private:
+    unsigned int m_version;
+    std::vector<unsigned char> m_program;
+
+public:
+    WitnessUnknown(unsigned int version, std::vector<unsigned char> program) : m_version(version), m_program(program) {}
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h]: Done

File: src/addresstype.h [hunk 11]
```diff
@@ -27,6 +27,20 @@ class CNoDestination {
...
+{
+private:
+    CPubKey m_pubkey;
+
+public:
+    PubKeyDestination(const CPubKey& pubkey) : m_pubkey(pubkey) {}
+
+    const CPubKey& GetPubKey() const { return m_pubkey; }
+
+    friend bool operator==(const PubKeyDestination &a, const PubKeyDestination &b) { return a.GetPubKey() == b.GetPubKey(); }
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h]: Done

File: src/addresstype.h (hunk 5 above)
Comment (achow101) [reply src/addresstype.h:74]: Done as suggested.

File: src/test/fuzz/util.cpp (hunk 6 above)
Comment (achow101) [reply src/test/fuzz/util.cpp]: Done

File: src/key_io.cpp [hunk 12]
```diff
@@ -190,7 +190,7 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
                 return CNoDestination();
             }
 
-            return WitnessUnknown{version, data};
+            return WitnessUnknown{static_cast<unsigned int>(version), data};
```
Comment (josibake) [inline src/key_io.cpp:193]: in https://github.com/bitcoin/bitcoin/pull/28246/commits/d33fa1fae66635992e6a0a980e7d16a9ebe4c852:

Is there a reason to do this here, as opposed to making an overloaded constructor for `WitnessUnknown`, like:

```
WitnessUnknown(int version, std::vector<unsigned char> program)
        : m_version(static_cast<unsigned int>(version)), m_program(std::move(program)) {}
```

that way, anything that creates a `WitnessUnknown` can just be `WitnessUnknown{version, program}`

**Review by josibake — COMMENTED** [review]
Looks good! I had some questions about the constructors for `CNoDestination` and `PubKeyDestination`, seems like there's an opportunity to make them consistent. Regarding one of the new comments, it mentions that `P2PK` is non-standard, but I don't think that's true.

As far as testing, I've been rebasing my silent payments PRs on this one and haven't ran into any issues

File: src/addresstype.h [hunk 13]
```diff
@@ -92,25 +100,30 @@ struct WitnessUnknown
...
+ *  * WitnessV0ScriptHash: TxoutType::WITNESS_V0_SCRIPTHASH destination (P2WSH address)
+ *  * WitnessV0KeyHash: TxoutType::WITNESS_V0_KEYHASH destination (P2WPKH address)
+ *  * WitnessV1Taproot: TxoutType::WITNESS_V1_TAPROOT destination (P2TR address)
+ *  * WitnessUnknown: TxoutType::WITNESS_UNKNOWN destination (P2W??? address)
  *  A CTxDestination is the internal data type encoded in a bitcoin address
  */
 using CTxDestination = std::variant<CNoDestination, PKHash, ScriptHash, WitnessV0ScriptHash, WitnessV0KeyHash, WitnessV1Taproot, WitnessUnknown>;
 
-/** Check whether a CTxDestination is a CNoDestination. */
+/** Check whether a CTxDestination corresponds to one with an address. */
```
Comment (S3RK) [inline src/addresstype.h:129]: IIUC `IsValidDestination` will return `true` for `PubKeyDestination`, but it doesn't have an address.

File: src/addresstype.h [hunk 14]
```diff
@@ -92,25 +100,30 @@ struct WitnessUnknown
...
 
 /**
- * Parse a standard scriptPubKey for the destination address. Assigns result to
- * the addressRet parameter and returns true if successful. Currently only works for P2PK,
- * P2PKH, P2SH, P2WPKH, and P2WSH scripts.
+ * Parse a scriptPubKey for the destination.
+ *
+ * For standard scripts that have addresses (and P2PK as an exception), a corresponding CTxDestination
+ * is assigned to addressRet.
+ * For all other scripts. addressRet is assigned as a CNoDestination containing the scriptPubKey.
```
Comment (S3RK) [inline src/addresstype.h:137]: The comment needs to be updated.
Below you say that `WitnessUnknown` is not standard, but it has its own type, not `CNoDestination`.

File: src/util/message.cpp [hunk 15]
```diff
@@ -47,7 +47,7 @@ MessageVerificationResult MessageVerify(
         return MessageVerificationResult::ERR_PUBKEY_NOT_RECOVERED;
     }
 
-    if (!(CTxDestination(PKHash(pubkey)) == destination)) {
+    if (!(PKHash(pubkey) == *std::get_if<PKHash>(&destination))) {
```
Comment (ismaelsadeeq) [inline src/util/message.cpp:50]: If IIUC this is a refactor.
So should be in it's own commit since the change does not concern 'Make WitnessUnknown members private' ?

File: src/addresstype.h (hunk 13 above)
Comment (ismaelsadeeq) [reply src/addresstype.h:129]: May be should be
```suggestion
/** Check whether a CTxDestination is not isCNoDestination . */
```
Since thats what it does

**Review by ismaelsadeeq — COMMENTED** [review]
Concept ACK

File: src/util/message.cpp (hunk 15 above)
Comment (josibake) [reply src/util/message.cpp:50]: see https://github.com/bitcoin/bitcoin/pull/28246#discussion_r1318874324

It's not actually a refactor, but a fix for some very strange / edge case behavior with `std::variant` and `std::vector`

File: src/addresstype.h [hunk 16]
```diff
@@ -14,9 +14,17 @@
...
 
 class CNoDestination {
+private:
+    CScript m_script;
+
 public:
-    friend bool operator==(const CNoDestination &a, const CNoDestination &b) { return true; }
-    friend bool operator<(const CNoDestination &a, const CNoDestination &b) { return true; }
+    CNoDestination() = default;
+    CNoDestination(const CScript& script) : m_script(script) {}
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h:22]: No, I think `WitnessUnknown` should actually be using a reference rather than value. AFAIK, both are equally efficient.

File: src/addresstype.h [hunk 17]
```diff
@@ -92,25 +100,30 @@ struct WitnessUnknown
...
- * A txout script template with a specific destination. It is either:
- *  * CNoDestination: no destination set
- *  * PKHash: TxoutType::PUBKEYHASH destination (P2PKH)
- *  * ScriptHash: TxoutType::SCRIPTHASH destination (P2SH)
- *  * WitnessV0ScriptHash: TxoutType::WITNESS_V0_SCRIPTHASH destination (P2WSH)
- *  * WitnessV0KeyHash: TxoutType::WITNESS_V0_KEYHASH destination (P2WPKH)
- *  * WitnessV1Taproot: TxoutType::WITNESS_V1_TAPROOT destination (P2TR)
- *  * WitnessUnknown: TxoutType::WITNESS_UNKNOWN destination (P2W???)
+ * A txout script categorized into standard templates.
+ *  * CNoDestination: A script, no corresponding address.
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h]: Clarified that it is optionally a script.

File: src/addresstype.h [hunk 18]
```diff
@@ -122,8 +136,8 @@ bool IsValidDestination(const CTxDestination& dest);
  * is assigned to addressRet.
  * For all other scripts. addressRet is assigned as a CNoDestination containing the scriptPubKey.
  *
- * Returns true for standard destinations - P2PK, P2PKH, P2SH, P2WPKH, P2WSH, and P2TR scripts.
- * Returns false for non-standard destinations - P2PK with invalid pubkeys, P2W???, bare multisig, null data, and nonstandard scripts.
+ * Returns true for standard destinations - P2PKH, P2SH, P2WPKH, P2WSH, and P2TR scripts.
+ * Returns false for non-standard destinations - P2PK, P2W???, bare multisig, null data, and nonstandard scripts.
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.h]: Clarified that it's looking for standard destinations with addresses. P2PK does not have an address.

File: src/rpc/output_script.cpp [hunk 19]
```diff
@@ -280,6 +280,11 @@ static RPCHelpMan deriveaddresses()
                 for (const CScript& script : scripts) {
                     CTxDestination dest;
                     if (!ExtractDestination(script, dest)) {
+                        // P2PK is no longer considered a valid destination
+                        // However combo will output P2PK and should just ignore that script
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/rpc/output_script.cpp]: Clarified that P2PK does not correspond to an address.

File: src/key_io.cpp (hunk 12 above)
Comment (achow101) [reply src/key_io.cpp]: Done

File: src/addresstype.h (hunk 13 above)
Comment (achow101) [reply src/addresstype.h:130]: The comment was what I intended for `IsValidDestination`, but the implementation got lost somewhere. Fixed.

File: src/addresstype.cpp [hunk 20]
```diff
@@ -108,6 +109,11 @@ class CScriptVisitor
         return dest.GetScript();
     }
 
+    CScript operator()(const PubKeyDestination& dest) const
+    {
+        return CScript() << ToByteVector(dest.GetPubKey()) << OP_CHECKSIG;
```
Comment (josibake) [inline src/addresstype.cpp:114]: in https://github.com/bitcoin/bitcoin/pull/28246/commits/22d54175ab14e6f8f0426dc015b897a4c4a35284:

This line is causing the fuzzer to fail in `src/test/fuzz/script.cpp:155`, because it first calls `GetScriptForDestination`, (which now returns a script for PubKeDestination), and then calls `IsValidDestination` and checks if the `dest.empty()` and `IsValidDestination` match. Can probably just special case `PubKeyDestination`? Or maybe there's a more clever way to handle it

Comment (josibake) [reply src/addresstype.cpp:114]: This should do the trick (and seems reasonable): https://github.com/bitcoin/bitcoin/commit/7431452ab61a46715c31c8e0453234d328973088#diff-23790a106f16f165ad579a3966ae9faa871f148ca4640bc870f6071a121d5314L156-R161

Comment (achow101) [reply src/addresstype.cpp:114]: Fixed

File: src/addresstype.h (hunk 5 above)
Comment (josibake) [reply src/addresstype.h:74]: in https://github.com/bitcoin/bitcoin/pull/28246/commits/c14ad74f8eeaa744d831a77654164c856a86fdee:

I'm getting the `-Wunitialized` error again, due to this line: 

https://github.com/bitcoin/bitcoin/blob/c14ad74f8eeaa744d831a77654164c856a86fdee/src/test/fuzz/key.cpp#L189

IIRC, this was fixed in a previous version of this PR, not sure when it changed, but this should be:

```
-        assert(CTxDestination{PKHash{pubkey}} == tx_destination);
+        assert(PKHash{pubkey} == *std::get_if<PKHash>(&tx_destination));
```

Comment (achow101) [reply src/addresstype.h:74]: Fixed in fuzz/key.cpp

**Comment by josibake** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/28246/commits/ad0c469d98c51931b98b7fd937c6ac3eeaed024e

Verified that the `-Wuninitialized` error is now fixed. Code looks good and I've been using this in https://github.com/bitcoin/bitcoin/pull/28122 and https://github.com/bitcoin/bitcoin/pull/28201

**Comment by Sjors** [comment]
> Verified that the `-Wuninitialized` error is now fixed.

Found another one post-merge (while testing things on master), although only when you `./configure` with `--enable-c++20`. It was introduced by 8dd067088d41f021b357d7db5fa5f0a9f61edddc. Compiled on Ubuntu 23.10 with gcc 12.3.0.

Log output: https://gist.github.com/Sjors/22873167e460eb1265652e5dca973db2

```
./configure --enable-c++20 --enable-suppress-external-warnings --disable-fuzz-binary --without-gui --enable-werror
```

**Comment by Sjors** [comment]
@MarcoFalke all I know is that `--enable-c++20` has worked fine on master, and most of the PRs I reviewed, for several months, but I have no idea whose fault it is :-) Compiler says No.

File: src/addresstype.cpp [hunk 21]
```diff
@@ -54,11 +54,12 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
...
-        if (!pubKey.IsValid())
-            return false;
-
-        addressRet = PKHash(pubKey);
-        return true;
+        if (!pubKey.IsValid()) {
+            addressRet = CNoDestination(scriptPubKey);
+        } else {
+            addressRet = PubKeyDestination(pubKey);
+        }
```
[the remark this answers is not in the document]
Comment (murchandamus) [reply src/addresstype.cpp:61]: Yes. See the first comment in this PR:

> `ExtractDestination`'s behavior is slightly changed for the above. It now returns `true` for those destinations that have addresses, so P2PK scripts now result in `false`. Even though it returns false for `CNoDestination`, the script will now be included in that `CNoDestination`.

[the remark this answers is not in the document]
Comment (denavila) [reply src/addresstype.cpp:61]: But why is P2PK treated differently now? 
Previously when the function returned "false", it meant the result is CNoDestination, but now that's ambiguous and could also be PubKeyDestination, and thus requires further checks. 
This kind of makes the return value not very useful, IMHO.

[the remark this answers is not in the document]
Comment (denavila) [reply src/addresstype.cpp:61]: Hmm, looks like OutputTypeFromDestination needs to be updated to support PubKeyDestination?
I'm hitting an issue in my PR, with some wallet tests that use addresses from mines blocks (GetScriptForRawPubKey(coinbaseKey.GetPubKey())), and OutputTypeFromDestination is returning nullopt now.

[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.cpp:61]: > Previously when the function returned "false", it meant the result is CNoDestination, but now that's ambiguous and could also be PubKeyDestination, and thus requires further checks.
> This kind of makes the return value not very useful, IMHO.

As noted in the change to the documentation for this function, `ExtractDestination` only returns true for things that have addresses. P2PK does not have an address. This is useful in other parts of the codebase and in future work such as silent payments.  If you want to check if it is a `CNoDestination`, then you can do that directly. However, it should generally be okay to not need to check for that. For the most part, checking for `CNoDestination` is not all that useful and there's probably a better way to achieve what you want to do.

> Hmm, looks like OutputTypeFromDestination needs to be updated to support PubKeyDestination?
> I'm hitting an issue in my PR, with some wallet tests that use addresses from mined blocks (GetScriptForRawPubKey(coinbaseKey.GetPubKey())), and OutputTypeFromDestination is returning nullopt now.

No, it does not. `OutputType` is also a synonym for address type in the context of giving an address to the user. P2PK does not have an address type and users cannot get an address for them. If you are using things that relied on the incorrect behavior of getting a `PKHash` for a P2PK script, then I suggest that you don't do that. P2PK is not `PKHash` and you shouldn't expect it to be giving you those scripts.

[the remark this answers is not in the document]
Comment (denavila) [reply src/addresstype.cpp:61]: Hmm, I don't think I understand why it's considered to have no address ...
In the wallet test, I'm using GetScriptForRawPubKey(coinbaseKey.GetPubKey()) to mine a block (via the MineBlock function), and then create a wallet with that coinbase key and issue transactions with coins returned from AvailableCoins.
With the new code path, it seems the address coming from AvailableCoins gets classified as PubKeyDestination, which fails to get an OutputTypeFromDestination, however before this change, this code worked and generated successful transactions.
Is there another way to mine blocks and get P2PK addresses from the coinbase key?

[the remark this answers is not in the document]
Comment (achow101) [reply src/addresstype.cpp:61]: > Hmm, I don't think I understand why it's considered to have no address

They literally do not have addresses. There is no address type which maps to a P2PK script. That this and other software sometimes display an address for P2PK scripts is simply wrong. They are displaying P2PK addresses. If you send to those addresses, a P2PK script is not created.

> In the wallet test, I'm using GetScriptForRawPubKey(coinbaseKey.GetPubKey()) to mine a block (via the MineBlock function), and then create a wallet with that coinbase key and issue transactions with coins returned from AvailableCoins.
> With the new code path, it seems the address coming from AvailableCoins gets classified as PubKeyDestination, which fails to get an OutputTypeFromDestination, however before this change, this code worked and generated successful transactions.

`AvailableCoins` does not operate on destinations, it operates on scripts. This is way more precise since there are more standard script types than there are standard addresses (e.g. P2PK, bare multisig). If you have modified it to operate on destinations, don't do that.

[the remark this answers is not in the document]
Comment (denavila) [reply src/addresstype.cpp:61]: Thanks! It's clear I didn't have a good understanding of the relationship between scripts and addresses. I'll move the discussion to my PR.

**Comment by Sjors** [comment]
@maflcko wrote:

> > Compiled on Ubuntu 23.10 with gcc 12.3.0.
> 
> Have you tried using 13.2 from the same distro? https://packages.ubuntu.com/mantic/g++

That did the trick. Although for some reason it picks `13.1.0` when I install `gcc-13` and `g++-13`.

File: src/wallet/wallet.cpp [hunk 22]
```diff
@@ -2213,15 +2213,13 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
...
+        } else if (std::get_if<WitnessV0KeyHash>(&recipient.dest)) {
             any_wpkh = true;
-        } else if (type == TxoutType::SCRIPTHASH) {
+        } else if (std::get_if<ScriptHash>(&recipient.dest)) {
             any_sh = true;
-        } else if (type == TxoutType::PUBKEYHASH) {
+        } else if (std::get_if<PKHash>(&recipient.dest)) {
             any_pkh = true;
         }
     }
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp:2225]: > Looking at the code, and the missing `explicit` keyword, I presume this broke this feature in the GUI?

Great catch. Can fix it with https://github.com/bitcoin/bitcoin/pull/28728#discussion_r1371706575.
